test/support/warning_filter.rb

```diff
@@ -0,0 +1,53 @@
+require 'pathname'
+require 'set'
+
+module RegexpParserTest
+  # Stream filter for warnings
+  class WarningFilter
+    PATTERN = /\A(?:.+):(?:\d+): warning: (?:.+)\n\z/
+
+    # Use warning filter during block execution
+    #
+    # @return [Array<String>]
+    def self.use
+      original = $stderr
+      $stderr = filter = new(original)
+      yield
+      filter.warnings
+    ensure
+      $stderr = original
+    end
+
+    # Initialize object
+    #
+    # @param [#write] target
+    #
+    # @return [undefined]
+    def initialize(target)
+      @target   = target
+      @warnings = Set.new
+    end
+
+    # Write message to target filtering warnings
+    #
+    # @param [String] message
+    #
+    # @return [self]
+    def write(message)
+      return target.write(message) if PATTERN !~ message
+
+      @warnings = warnings + Set.new([message.chomp])
```

Comment:
Also what you're doing here is still mutation. You're replacing the state of `@warnings` in the object with a new object. It'd be different if you had a fully immutable object and you have `#write` return a totally new object, but since this is wrapping a standard IO object you have to adhere to it's API so that's not an option.